src/ngp_algorithms/NodalGradPOpenBoundaryAlg.C

```diff
@@ -0,0 +1,162 @@
+/*------------------------------------------------------------------------*/
+
+#include "ngp_algorithms/NodalGradPOpenBoundaryAlg.h"
+#include "Algorithm.h"
+
+#include "BuildTemplates.h"
+#include "master_element/MasterElement.h"
+#include "master_element/MasterElementFactory.h"
+#include "ngp_algorithms/ViewHelper.h"
+#include "ngp_utils/NgpLoopUtils.h"
+#include "ngp_utils/NgpFieldOps.h"
+#include "Realm.h"
+#include "ScratchViews.h"
+#include "SolutionOptions.h"
+#include "utils/StkHelpers.h"
+
+namespace sierra{
+namespace nalu{
+
+//==========================================================================
+// Class Definition
+//==========================================================================
+// NodalGradPOpenBoundary - adds in boundary contribution
+//                                      for elem/edge proj nodal gradient
+//==========================================================================
+//--------------------------------------------------------------------------
+//-------- constructor -----------------------------------------------------
+//--------------------------------------------------------------------------
+template <typename AlgTraits>
+NodalGradPOpenBoundary<AlgTraits>::NodalGradPOpenBoundary(
+  Realm &realm,
+  stk::mesh::Part *part,
+  const bool useShifted)
+  : Algorithm(realm, part),
+    useShifted_(useShifted),
+    zeroGrad_(realm_.solutionOptions_->explicitlyZeroOpenPressureGradient_),
+    massCorr_(realm_.solutionOptions_->activateOpenMdotCorrection_),
+    exposedAreaVec_ (get_field_ordinal(realm_.meta_data(), "exposed_area_vector", realm.meta_data().side_rank())),
+    dualNodalVol_   (get_field_ordinal(realm_.meta_data(), "dual_nodal_volume")),
+    exposedPressureField_   (get_field_ordinal(realm_.meta_data(), (massCorr_ ? "pressure" : "pressure_bc"))),
+    pressureField_ (get_field_ordinal(realm_.meta_data(), "pressure")),
+    gradP_         (get_field_ordinal(realm_.meta_data(), "dpdx")),
+    coordinates_   (get_field_ordinal(realm_.meta_data(), realm.get_coordinates_name())),
+    meFC_  (MasterElementRepo::get_surface_master_element<typename AlgTraits::FaceTraits>()),
+    meSCS_ (MasterElementRepo::get_surface_master_element<typename AlgTraits::ElemTraits>()),
+    faceData_(realm.meta_data()),
+    elemData_(realm.meta_data())
+{
+  const int coordinates = get_field_ordinal(realm.meta_data(), realm.get_coordinates_name());
+  faceData_.add_coordinates_field(coordinates, AlgTraits::nDim_, CURRENT_COORDINATES);
+  elemData_.add_coordinates_field(coordinates, AlgTraits::nDim_, CURRENT_COORDINATES);
+  faceData_.add_cvfem_face_me   (meFC_);
+  elemData_.add_cvfem_surface_me(meSCS_);
+
+  faceData_.add_face_field(exposedAreaVec_, AlgTraits::numFaceIp_, AlgTraits::nDim_);
+  faceData_.add_gathered_nodal_field(dualNodalVol_,1);
+  faceData_.add_gathered_nodal_field(exposedPressureField_,1);
+  faceData_.add_gathered_nodal_field(pressureField_,1);
```

Comment:
This should be registered to `elemData_`, because we will be using it to access nodes of opposingFace